File: src/main/java/com/github/unldenis/hologram/line/BlockLine.java
package com.github.unldenis.hologram.line;

import com.github.unldenis.hologram.packet.PacketContainerSendable;
import com.github.unldenis.hologram.packet.PacketsFactory;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.Plugin;

public final class BlockLine implements ILine<ItemStack> {

  private final Line line;
  private final PacketContainerSendable entityMetadataPacket;

  private ItemStack obj;

  public BlockLine(Line line, ItemStack obj) {
    if (!obj.getType().isBlock()) {
      throw new UnsupportedOperationException(
          "'%s' is not a block. Are you looking for the new experimental ItemLine class?".formatted(obj.getType().name()));
    }
    this.line = line;
    this.entityMetadataPacket = PacketsFactory.get().metadataPacket(line.getEntityID(), null);

    this.obj = obj;
  }

  @Override
  public Plugin getPlugin() {
    return line.getPlugin();
  }

  @Override
  public Type getType() {
    return Type.BLOCK_LINE;
  }

  @Override
  public int getEntityId() {
    return line.getEntityID();
  }

  @Override
  public Location getLocation() {
    return line.getLocation();
  }

  @Override
  public void setLocation(Location location) {
    line.setLocation(location);
  }

  @Override
  public ItemStack getObj() {
    return obj.clone();
  }

  @Override
  public void setObj(ItemStack obj) {
    this.obj = obj;
  }

  @Override
  public void hide(Player player) {
    line.destroy(player);
  }

  @Override
  public void teleport(Player player) {
    line.teleport(player);
  }

  @Override
  public void show(Player player) {
    line.spawn(player);
    entityMetadataPacket.send(player);
    this.update(player);
  }

  @Override
  public void update(Player player) {
    PacketsFactory.get()
        .equipmentPacket(line.getEntityID(), this.obj, false)
        .send(player);
  }

}

File: src/main/kotlin/com/github/unldenis/hologram/line/BlockLine.kt
package com.github.unldenis.hologram.line

import com.comphenix.protocol.events.PacketContainer
import com.github.unldenis.hologram.ext.send
import com.github.unldenis.hologram.packet.PacketsFactory
import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.Plugin

class BlockLine(line: Line, override var obj: ItemStack) : ILine<ItemStack> {
    private val line: Line
    private val entityMetadataPacket: PacketContainer
    override val plugin: Plugin
        get() = line.plugin

    override val type: ILine.Type
        get() = ILine.Type.BLOCK_LINE

    override val entityId: Int
        get() = line.entityID

    override val location: Location?
        get() = line.location

    init {
        if (!obj.type.isBlock) {
            throw UnsupportedOperationException(
                "'%s' is not a block. Are you looking for the new experimental ItemLine class?".formatted(obj.type.name)
            )
        }
        this.line = line
        this.entityMetadataPacket = PacketsFactory.get().metadataPacket(line.entityID, null)
    }


    override fun setLocation(value: Location) {
        line.location = value
    }

    override fun hide(player: Player) {
        line.destroy(player)
    }

    override fun teleport(player: Player) {
        line.teleport(player)
    }

    override fun show(player: Player) {
        line.spawn(player)
        entityMetadataPacket.send(player)
        this.update(player)
    }

    override fun update(player: Player) {
        PacketsFactory.get()
            .equipmentPacket(line.entityID, this.obj, false)
            .send(player)
    }
}
